lib/Parse/ParseDecl.cpp

```diff
@@ -300,18 +300,250 @@ getStringLiteralIfNotInterpolated(Parser &P, SourceLoc Loc, const Token &Tok,
                                                  Segments.front().Length));
 }
 
-bool Parser::parseSpecializeAttribute(swift::tok ClosingBrace, SourceLoc AtLoc,
-                                      SourceLoc Loc, SpecializeAttr *&Attr) {
-  assert(ClosingBrace == tok::r_paren || ClosingBrace == tok::r_square);
-  SourceLoc lParenLoc = consumeToken();
-  bool DiscardAttribute = false;
-  StringRef AttrName = "_specialize";
+AvailableAttr *Parser::parseExtendedAvailabilitySpecList(
+    SourceLoc AtLoc, SourceLoc AttrStartLoc, StringRef Platform,
+    StringRef AttrName, SourceLoc AttrLoc, bool &DiscardAttribute) {
```

Comment:
The motivation for factoring out this function was that it could create a `SyntaxParsingContext` with kind `AvailabilitySpecList` for its entire scope so we don't have to create a code block that for that context that spans nearly 200 lines of code. That's why it also doesn't parse the trailing `)`. 
I will implement 2, 3, and 4. Thanks.

Reply:
Have you tried `collectNodeInPlace(SyntaxKind::AvailabilitySpecList)` without creating `SyntaxParsingContext` in `parseExtendedAvailabilitySpecList`?

Reply:
Actually, `AtLoc` and`AttrLoc` are different, `AtLoc` refers to where the `@` is, `AttrLoc` where the name of the attribute starts.

Reply:
Yes, that doesn't work because the opening `(` is still on the stack of items to be parsed and doesn't fit into the `AvailabilitySpecList`

Reply:
`AtLoc` != `AttrLoc` but `AttrStartLoc` == `AttrLoc`, right?

Reply:
Yes. Sorry, misread your comment. You're right of course

Reply:
> Yes, that doesn't work because the opening `(` is still on the stack of items

It should work. For example, in `parseDeclEnumCase()`, `collectNodesInPlace(SyntaxKind::EnumCaseElementList)` turns
`{ 'case', element, element, element ...}` into `{ 'case', element-list }`.

Reply:
OK, it does seem to work. I'm pretty sure I tried before and it didn't but maybe I'm mixing things up. It's not needed though since I keep the parsing logic for `)` outside of the refactored function.